### ISSUE TITLE
gnovm/pkg/gnolang: hoist out never changing maximums and reuse them in comparisons in bit shifting result overflow detection, much faster detection

### ISSUE THREAD

**Issue by odeke-em** [gnolang/gno#3194]
## Problems
- [ ] The current overflow detection involves performing the actual bit shifts on the big.Int values then comparing against maximums. However, that code re-allocates never changing values and moreover into big.Int which is unnecessarily expensive for every single call
![IMG_2329](https://github.com/user-attachments/assets/46276e4e-7fcd-4b61-b6cd-17118202b6e4)
- [ ] while there is some overflow detection that code doesn’t check for underflow which will happen for unsigned variants for example `(1<<63)<<2` won’t report a problem 
- [ ] The code as written seems to do the opposite of the expectation, checkOverflow returns false in the correct cases when there is no overflow meanwhile the panic as written will occur when that code doesn’t overflow 
![IMG_2330](https://github.com/user-attachments/assets/1f51004f-68e9-4281-91ad-8a18674326c6)

## Finding
On my 6AM code audit, I examined this commit https://github.com/gnolang/gno/commit/f27b182702d278d79d8b1f95d06913a20f4f5f62 we can see that there is some overflow detection here
https://github.com/gnolang/gno/blob/f27b182702d278d79d8b1f95d06913a20f4f5f62/gnovm/pkg/gnolang/op_binary.go#L1217
https://github.com/gnolang/gno/blob/f27b182702d278d79d8b1f95d06913a20f4f5f62/gnovm/pkg/gnolang/op_binary.go#L1226
https://github.com/gnolang/gno/blob/f27b182702d278d79d8b1f95d06913a20f4f5f62/gnovm/pkg/gnolang/op_binary.go#L1235
https://github.com/gnolang/gno/blob/f27b182702d278d79d8b1f95d06913a20f4f5f62/gnovm/pkg/gnolang/op_binary.go#L1244
https://github.com/gnolang/gno/blob/f27b182702d278d79d8b1f95d06913a20f4f5f62/gnovm/pkg/gnolang/op_binary.go#L1253
https://github.com/gnolang/gno/blob/f27b182702d278d79d8b1f95d06913a20f4f5f62/gnovm/pkg/gnolang/op_binary.go#L1262
https://github.com/gnolang/gno/blob/f27b182702d278d79d8b1f95d06913a20f4f5f62/gnovm/pkg/gnolang/op_binary.go#L1271
https://github.com/gnolang/gno/blob/f27b182702d278d79d8b1f95d06913a20f4f5f62/gnovm/pkg/gnolang/op_binary.go#L1280
https://github.com/gnolang/gno/blob/f27b182702d278d79d8b1f95d06913a20f4f5f62/gnovm/pkg/gnolang/op_binary.go#L1289
https://github.com/gnolang/gno/blob/f27b182702d278d79d8b1f95d06913a20f4f5f62/gnovm/pkg/gnolang/op_binary.go#L1298
https://github.com/gnolang/gno/blob/f27b182702d278d79d8b1f95d06913a20f4f5f62/gnovm/pkg/gnolang/op_binary.go#L1307

## Suggestions/fixes
- [ ] Please make those into global variables that get reused and can gut out expensive and unnecessary re-allocations
- [ ] Checking for overflows can be trivially performed by noticing that a left shit will always move the most significant bit to the left, hence if the bit shift value is not 0, simply examine the number of shifts in rv and those in lv, if rv+bits(lv) > bits(max*int), then likely to have an overflow: `5+bits(1<<62) > bits(math.MaxUint64)`

Kindly /cc-ing @jaekwon @petar-dambovaliev @ltzmaxwell

**Comment by petar-dambovaliev** [comment]
Notice that this code runs only for preprocessing for constant expressions.
It does not run during runtime.

**Comment by odeke-em** [comment]
@petar-dambovaliev thanks for your response, but does that remove the facts raised in the issues? I see you've closed the issues. That can be super confusing for auditors and other developers.

**Comment by petar-dambovaliev** [comment]
> @petar-dambovaliev thanks for your response, but does that remove the facts raised in the issues? I see you've closed the issues. That can be super confusing for auditors and other developers.

It seems, you did not understand my response. Yes, this can be made faster but i don't see the point of it as it runs only once.

**Comment by odeke-em** [comment]
There also seem to be legitimate other bugs in that code that I had raised like `r.Cmp(big.NewInt(math.MaxInt64)) != 1` will always return false for values that haven't overflowed but your code does the opposite per https://go.dev/play/p/p4eFKCHOTq8
```go
package main

import (
	"math"
	"math/big"
)

var bigMaxInt64 = big.NewInt(math.MaxInt64)

func cmp(a int64) bool {
	return big.NewInt(a).Cmp(bigMaxInt64) != 1
}

func main() {
	println(cmp(10))
}
```

I am not trying to be confrontational but I don't understand why this issue has been closed twice yet it has underflow issues, the logic seems to do the opposite hence seems untested, the other issues still linger. How are bugs followed up in the Gno ecosystem? I am trying to be a contributor and for me the usual way I've worked on other projects is that you raise issues so that there is a reference attached to the fixes.
